api/provider/aws/deploy_create.go

```diff
@@ -0,0 +1,136 @@
+	familyName := fqDeployID
+
+	taskDefinition, err := d.renderTaskDefinition(req.DeployFile, fqDeployID)
+	if err != nil {
+		return nil, err
+	}
+
+	input := &ecs.TaskDefinition{}
+	input.SetFamily(familyName)
+	input.SetTaskRoleArn(aws.StringValue(taskDefinition.TaskRoleArn))
+	input.SetNetworkMode(aws.StringValue(taskDefinition.NetworkMode))
+	input.SetContainerDefinitions(taskDefinition.ContainerDefinitions)
+	input.SetVolumes(taskDefinition.Volumes)
+	input.SetPlacementConstraints(taskDefinition.PlacementConstraints)
+
+	taskDefinitionOutput, err := d.createTaskDefinition(input)
+	if err != nil {
+		return nil, err
+	}
+
+	bytes, err := json.Marshal(taskDefinitionOutput)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to extract deploy file: %s", err.Error())
+	}
+
+	deploy := &models.Deploy{
+		// TODO: we previously removed prefix from deployID ala below. Is this
+		// no longer needed?
+		// func removePrefix(id string) string {
+		// 	return strings.TrimPrefix(id, PREFIX)
+		// }
+		DeployID:   deployID,
+		Version:    d.getDeployRevision(deployID),
+		DeployFile: bytes,
+	}
+
+	if err := d.TagStore.Insert(models.Tag{EntityID: deploy.DeployID, EntityType: "deploy", Key: "name", Value: req.DeployName}); err != nil {
+		return deploy, err
+	}
+
+	if err := d.TagStore.Insert(models.Tag{EntityID: deploy.DeployID, EntityType: "deploy", Key: "version", Value: deploy.Version}); err != nil {
+		return deploy, err
+	}
+
+	return deploy, nil
+}
+
+func (d *DeployProvider) createTaskDefinition(taskDefinitionRequest *ecs.TaskDefinition) (*ecs.TaskDefinition, error) {
+
+	input := &ecs.RegisterTaskDefinitionInput{}
+	input.SetFamily(aws.StringValue(taskDefinitionRequest.Family))
+	input.SetTaskRoleArn(aws.StringValue(taskDefinitionRequest.TaskRoleArn))
+	input.SetNetworkMode(aws.StringValue(taskDefinitionRequest.NetworkMode))
+	input.SetContainerDefinitions(taskDefinitionRequest.ContainerDefinitions)
+	input.SetVolumes(taskDefinitionRequest.Volumes)
+	input.SetPlacementConstraints(taskDefinitionRequest.PlacementConstraints)
+
+	if err := input.Validate(); err != nil {
+		return nil, err
+	}
+
+	_, err := d.AWS.ECS.RegisterTaskDefinition(input)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: return original request or RegisterTaskDefinition() input?
+	return taskDefinitionRequest, nil
+}
+
+func (d *DeployProvider) renderTaskDefinition(body []byte, familyName string) (*ecs.TaskDefinition, error) {
+	var taskDefinition *ecs.TaskDefinition
+
+	if err := json.Unmarshal(body, &taskDefinition); err != nil {
+		return nil, fmt.Errorf("Failed to decode deploy: %s", err.Error())
+	}
+
+	if len(taskDefinition.ContainerDefinitions) == 0 {
+		return nil, fmt.Errorf("Deploy must have at least one container definition")
+	}
+
+	if taskDefinition.Family != nil && taskDefinition.Family != aws.String(familyName) {
+		return nil, fmt.Errorf("Custom family names are currently unsupported in Layer0")
```

Comment:
Use our own error type: `errors.New(errors.InvalidRequest, "Custom family names are unsupported in Layer0")`

Reply:
What's the criteria for using our own error type versus `fmt.Errorf`?

Reply:
If we can capture an error and assign a meaningful error code to it